tokio/src/runtime/task/harness.rs

```diff
@@ -313,9 +314,17 @@ where
 
         let snapshot = self.state().transition_to_complete();
 
-        // We catch panics here in case dropping the future or waking the
+        // We catch panics here in case dropping the future, invoking a hook or waking the
         // JoinHandle panics.
         let _ = panic::catch_unwind(panic::AssertUnwindSafe(|| {
+            if let Some(f) = self.trailer().hooks.task_terminate_callback.as_ref() {
+                f(&TaskMeta {
+                    #[cfg(tokio_unstable)]
+                    id: self.core().task_id,
+                    _phantom: Default::default(),
+                })
+            }
```

Comment:
Does it make more sense to place it below the `drop_future_or_output`/`wake_join` calls? That way, the task will appear finished to the hook.

Reply:
Done!